### DISCUSSION TITLE
Some feedbacks about the O3DE building process.

### DISCUSSION THREAD

**Discussion by AndreaCatania** [o3de/o3de#3943] [Ideas]
Hello, In these days I been checking O3DE, and while there are a lot of things that I'll have to check, I'm already able to give some constructive feedbacks for the building process.

Building is massive; de-bloat the building process is necessary if you really expect that people will contribute to the engine at big scale. Here the list of things, that doesn't work so nice right now:

**Build approaches.**
In O3DE, there are three build approaches:
- Project-Centric - Engine SDK
- Project-Centric - Engine Source
- Engine-Centric

In my opinion there are too many building approaches, that just make things more confusing. I think the `Engine-Centric` approach is not so useful, since you can already fix or implement new engine features, by using the `Project-Centric - Engine Source` approach.
The `Project-Centric - Engine Source` approach, is really useful for an Engineer working on a game, since he can immediately debug the engine to fix a bug, or implement a new feature at core level, and customize O3DE as needed by the game. At the same time, an O3DE contributor, can just work on a new feature, using its workbench test project.

Removing the approach of `Engine-Centric`, will automatically make the `Project-Centric - Engine Source` adopted by anyone, with the following benefits:
- Documentation will be simpler (you have 1 feature less to document).
- Dev Engine Contributors will use the exact same approach used by Engine Dev Studios to customize the engine: this will simplify things.
- Less code to support, Less bugs to fix.

**Compilation size**
I'm a linux user, but I wanted to give a try to the engine, so I booted my Windows partition, where I've around **70GB** free space available. I was unable to build the engine and test it, because the building size is insane and the compiler needed even more space to finish the compilation. I didn't even have the chance to generate the project files.

Have so bloated compiled code, make things quite difficult also when you have to share the editor to your team, that has to download an enormous amount of data each time a new feature is available.

I don't know why the engine has to take so much space, but clearly this is an issue. Godot is really minimal, and while I don't expect that O3DE would take the same size, I also don't expect that just the engine would fill 70GB. Especially if we consider that this engine is advertised as modular.

In my opinion, the core should be minimal so that it compiles fast and doesn't fill your SSD for no reason.

**Python**
This part is weird, why the engine needs a custom python version? If it doesn't need a custom version, it would be much better to insert python as dependency, so the user can decide how to download it. 

**Packages**
In my opinion, the packages should be compiled locally by the gem that uses it or by the core, instead to be downloaded from somewhere already compiled.
The issue with this approach is that, if there is a bug inside thirdparty library (let's say PhysX), it's not possible to just open the `.cpp` file and fix it. However, in general I think it's a much solid solution let the compiler, compile all the thirdparty libraries.

**Engine installation**
Install the engine, while you chose to build the engine source is quite weird. Usually, make sense to install the engine, if you use a precompiled SDK that you download from internet, but when you build the engine, the compiler should just produce a binary that you can use / share to your team, etc..

**The binaries are too spreaded**
I'm not 100% sure if it's already possible to produce a single directory that contains the engine + editor + the projects shared libraries, so that you can easily provide the editor to your art team. So far it seems that all the files are spreaded between o3de-packages, o3de, and the project.

A separate command that would package it, to use only on the CI, would be enough I guess.

**One step building**
As is now, to build the project, it's needed to execute something like 7 complex commands. This gives a lot of room for mistakes.
In my opinion it should look like:
1. `git clone git@github.com:AndreaCatania/o3de.git`
2. `o3de/scripts/o3de.bat --generate-project my-test-project`
3. `cd my-test-project`
4. `cmake -B build/vs_2019`
5. `cmake --build build/vs_2019 --target release`

The above steps, should also build the engine just cloned and the project. I should be able to open the project using [visual studio code](https://code.visualstudio.com/), and be able to modify anything: core, gems, third party: This is the `Project-Centric - Engine Source` approach

# Conclusion
I'm quite confident that it's possible to rethink a bit the building process to simplify it, so to reduce drastically the binary size and make it much more easy to setup a CI to share the editor and game binary to the team.

However, I like some concepts of this engine, and I think that we can work toward improving it. Thanks for your effort making this!

**Comment by nemerle** [comment]
Just a single comment for now:
`In my opinion, the packages should be compiled locally by the gem that uses it or by the core, instead to be downloaded from somewhere already compiled.`

There are two problems with that:
* Build time increase ( building a full Qt5 during full engine/editor build would be painful )
* Compilation of a dependency might require additional dependencies - thus complicating things.

**Reply by AndreaCatania** [reply]
This is a good point, and as part of the fix, we should consider to remove bloated libraries, like Qt5 and use much simpler one. The best would be use ImGUI also for the editor, so that the editor and the game would use the same UI mechanisms, which is a good idea in any case.

**Reply by amznestebanpapp** [reply]
ImGUI and Qt are quite different (e.g. ImGUI is stateless, Qt is statefull, ImGUI is an overlay-UI, Qt uses the OS's window system). That replacement may not be practical, its widely used and the stateless nature may affect performance...
Reducing the size of Qt is possible. E.g. we have qtquick, webengine, qml and other libraries, qtcreator and other tools that are not usually being used from the package.

Note Qt is an editor-time dependency, not a game/server runtime dependency. Meaning, your game should not need Qt to run.

**Reply by AndreaCatania** [reply]
Yup, though if turns out that keep Qt means have a giant library that bloats the editor compilation and however the engine complexity, at some point we should really think about replacing it with something in house made, or with another lib that works better.

Even if Qt is just an in Editor dependency that doesn't affect the final game, it affects the ability to use the engine; making the engine less enjoyable to use and so to contribute to. The only thrust for a contributor to actually contribute to an open source engine is the enjoyment of it, and the building is the most fundamental piece of the software that influence that.

The building size is insane, and I think O3DE has many dependencies that are bloating the engine more than the benefit they are providing. For example the boost library.

I hope we will setup a plan to fix it, I would really like to see this happen and I hope you don't underestimate the importance of making the engine ergonomic to use for its first user class (the devs that makes games and contribute to the engine).

**Reply by nemerle** [reply]
I'd say we need to look at the current causes of bloat first since Qt itself does not cause any of it at the moment :smile: 

Currently used Qt libraries take only ~30Mb ( when debug symbols are stripped from them )

**Reply by amznestebanpapp** [reply]
The engine and the editor do not depend on boost. We have a 3p that depends on it (OpenImageIO), but all that is pre-compiled, even when building from source. We dont even have boost as a 3rdParty that you pull as a dependency.

Qt itself doesnt add that much to compilation, you can run build insights or some other build profiler to validate. There are multiple places where Qt is misused (e.g. we include `QtWidgets` instead of including specific header files, or headers are being included from headers instead of forward-declaring). Those misuses are not entirely qt's fault (the existence of those "big headers" could be argued). 
The size of the precompiled version of Qt barely affects build time (it actually just affects linking time in a small measurement). Since we dont link against a lot of qt, but specific libraries, the amount is really not meaningful.

I agree that 6.52Gb on Qt's 3rdParty is not ideal. There is a lot that can be removed from there. But removing things from there will not improve build time, it will just reduce the download size of that 3p prebuilt package (which is definitely a good thing to improve). 

However, build time can be reduced by removing UIs that are not used, we have a good amount of Legacy stuff in the EditorLib.

**Comment by nemerle** [comment]
My opinion on Qt: I would not remove it, since it provides a boatload of features that are not available in ImGui without a huge amount of additional work. Going even further, I'd love to have QtQuick UI  gem :smile:

**Reply by amznestebanpapp** [reply]
we actually removed any use of QtQuick because it brings a lot of undesired dependencies to the editor. QtQuick brings a whole web engine as a dependency. It also requires a special deploy of libraries to work (there are non-obvious runtime dependencies that need to be added).
Adding it as a gem and putting the dependencies there is possible and probably a better approach that having it available across the engine.

**Comment by AndreaCatania** [comment]
Yee, it's probably a lot of work I agree, and it's not the bigger stop, so it's fine do it late. On the bright side however that change would also provide some UI features to use in game, other than debloat the engine :D

**Comment by amznestebanpapp** [comment]
> So far it seems that all the files are spreaded between o3de-packages, o3de, and the project.

- The files in o3de-packages are needed for building the engine. Once the engine is built, those are not needed for running. All runtime dependencies are copied to the output.
- The files in o3de are needed for building the engine. Once the engine is built and installed, you dont need them to build a project (project-centric engine-sdk).
- For a project, we dont have a way to distribute a prebuilt version of the project to artists. Today, the expectation is that you distribute the prebuilt version of the engine and the artists build the project. The reason for this is that the artist will want to edit the gems that are enabled, edit assets, build levels, etc. Some of those things depend on the project's build. There are ways to avoid most of the build part, but you need to know exactly how to enable those things manually. For example: the list of gems that gets loaded at runtime are generated by cmake configure. If you construct manually a setreg file, you can avoid building. Another example: to build monolithically, the game has to link statically to enabled gems. This means that enabling/disabling a gem is no longer a runtime-only decision, it needs to rebuild.
We hope to enable a workflow where you can provide a precompiled version of the SDK+project to an artist (within certain limitations). But we dont have it today.

IF you are using project-centric engine-source, then you do need `o3de` and `o3de-packages` because you are building the engine as well. 

> A separate command that would package it, to use only on the CI, would be enough I guess.

This is done with `cmake install`. There is also `cpack` which is implemented in Windows and produces the installer.
The first one (`cmake install`) is what would produce a "relocatable" folder (this is what is used to produce the SDK).
We do validate this in CI, you can check the CI scripts in `scripts/build`

> Engine installation
> Install the engine, while you chose to build the engine source is quite weird. Usually, make sense to install the engine, if you use a precompiled SDK that you download from internet, but when you build the engine, the compiler should just produce a binary that you can use / share to your team, etc..

In this case you don't need to "install the engine". You can either register it or you can instruct your project where the engine is.
What is needed is for the project to know which engine you want to use and where that is. Which engine you want to use is defined in your `project.json` file (is just a "name"). Where that engine is is defined in your `~/.o3de/o3de_manifest.json`). You can also pass the location of the engine with `CMAKE_MODULE_PATH` to your project. The engine has a name defined in `engine.json`, that is used for the registration and is what should match. When the engine is installed, the name of the engine is different so you can choose which one you want to use from the project.

> `cmake --build build/vs_2019 --target release`

release is a configuration, not a target. 
If you are just passing the release configuration without indicating a target, this could be part of the problem. That would instruct to build everything. Since we want to be able to select the gems you want to load in your project at runtime and we also want for "build everything" to mean "build everything", you may be building way more than what you need. You should be building targets like `Editor` or your `<project>.GameLauncher`.
Building everything also builds tests, not-frequently-used tools, not-enabled gems, etc.
Build size should be improved by building only what you need. You can also e.g. remove tests by passing `LY_DISABLE_TEST_MODULES` (but dont need to if you are building what you need, e.g. the `Editor` doesnt depend on tests)
Can you confirm what command you used for generation and which one for building?

Really appreciate the feedback. Wanted to let you know that your concerns are taken into account. Some of this is being worked on:
- there is cleanup that is continuously happening (@nemerle has done several ones)
- there is ongoing improvements to "O3DE as SDK". For example, yesterday I merged a change that enables to generate a `cmake install` for a project. This allows to do something like a "release monolithic" layout of a game. Still not optimized, its generating a pak with everything, but its a starting point.

I would like to get some more feedback/help on identifying practical short/mid term solutions out of your analysis. So far, I identified:
- remove unnecessary bloat from Qt's package
- make sure that we have good documentation around the build time impact that "building everything" vs "building specific targets" would have
- reduce the amount of assets we have in o3de, currently my o3de folder (without .git) takes 7Gb, `AutomatedTesting` takes 2.7Gb. As a workaround you can delete (or do a sparse-checkout somehow) the "AutomatedTesting" folder, that will remove those 2.7Gb of the `o3de` folder.

**Reply by AndreaCatania** [reply]
Nice, I'm glad you are working on reducing compilation size and complexity!

> Build size should be improved by building only what you need.

Be able to customize the building is great, and it's the first feature I would expect for a modular engine. However, I feel like that the size issue is hidden somewhere else, and be able to chose what to compile is not a solution to fix it.

> there is ongoing improvements to "O3DE as SDK". 

I would like to contribute to the engine, and use the SDK would not help me, neither other contributors.

> reduce the amount of assets we have in o3de, currently my o3de folder (without .git) takes 7Gb, AutomatedTesting takes 2.7Gb. As a workaround you can delete (or do a sparse-checkout somehow) the "AutomatedTesting" folder, that will remove those 2.7Gb of the o3de folder.

O3DE was around 20GB just before starting the build, so it keep growing by 50GB during the building, and it stop because I didn't have enough space. If I understood correctly, it takes around 130GB of space, so remove 10GB of stuff is great, but doesn't fully solve the issue.

Sorry, I don't want to sound pedantic; I just want to put the focus on the right direction. The editor size is important, since O3DE is not the only project I use on my PC. It's a new project and if it fills my SSD it's problematic. 

I've removed O3DE, since it was occupying all my SSD. I'll try to test it again in few days and take a more detailed info.

**Comment by amznestebanpapp** [comment]
> O3DE was around 20GB just before starting the build, so it keep growing by 50GB during the building, and it stop because I didn't have enough space. If I understood correctly, it takes around 130GB of space, so remove 10GB of stuff is great, but doesn't fully solve the issue.

Really curious where those 20Gb pre-build are coming from. Locally, my o3de folder without the `build` folder is 7Gb (including `.git`).
My build folder is ~100Gb and I have debug/profile/release. Just profile should be less than half. So you may get to 50Gb overall, but you should not have started with 20Gb.
Are you cloning other repositories besides `o3de`? for example, are you also cloning `o3de-atom-sampleviewer`?
For building, just to confirm, you are using `LY_UNITY_BUILD`, right?

**Comment by Clockwork-Muse** [comment]
> This part is weird, why the engine needs a custom python version? If it doesn't need a custom version, it would be much better to insert python as dependency, so the user can decide how to download it.

Because you want a pinned python version.  Especially since this engine is also targeting VFX workflows, which have a standardized python version.
